### PR TITLE
Integrate ASP.NET API Versioning

### PR DESCRIPTION
**Pull request by commonsensesoftware**
## Overview

At the request of #21, this PR adds support for API Versioning to the reference application.

## Additions

- APIs support versions
- All API versions are explicit
- The existing, default API version `1.0` is retained
  - It might be more practical to switch from the numeric to the date format
- API Explorer extensions have been added to discover and include versioned API information
- Swashbuckle customization has been included to leverage all of the versioned API information
- `HttpClient` extensions are included to integrate the target API version in outgoing requests

## Changes

The application fundamentally changes from versioning by URL segment to versioning by query string parameter. Without ASP.NET Versioning in place, I understand why that approach was employed. Versioning by URL segment, despite being very popular, is not RESTful; it violates the _Uniform Interface_ constraint. A resource can have multiple identifiers, but that is not what `v1/catalog/item/42` versus `v2/catalog/item/42` implies. The expectation is a different representation versus a different resource. Representations are indicated by a media type in `Accept` or `Content-Type`. A query string parameter is a pragmatic balance. It doesn't change the resource identifier (e.g. the _path_), but is still _easy_ to indicate API versions without having to set header values.

This is the just the tip of the iceberg of problems that can arise for versioning by URL segment. HATEOAS becomes nearly impossible or coupled as APIs evolve independently over time. In a self-contained application, where both sides are owned, that might not be an issue. If this application were to hypothetically open up to external callers (ex: for consignment), there could be challenges when linking resources. As an example, if `order/abc` had an item with a link to `catalog/item/42` and there are two versions, how would the server know which version to pick? It might assume the same version as the service itself or the incoming request, but those are both flawed. There is no requirement or guarantee a client uses the same version across all APIs. This _can_ be facilitated by enforcing _symmetric versioning_ across all APIs, but that requires adding new API versions when no changes have occurred. Ultimately, the client should **always** explicitly specify the API version they want (much in the same way that `Accept` indicates what format they want it in).

As a reference application, this example should provide a template that leads developers to the _Pit of Success_. Other arguments aside, switching to the query string approach has the following additional benefits:

- Alignment to how [Azure APIs are versioned](https://github.com/microsoft/api-guidelines/blob/vNext/azure/Guidelines.md)
- Alignment to the recommended [API Versioning examples](https://github.com/dotnet/aspnet-api-versioning/tree/main/examples/AspNetCore/WebApi/MinimalOpenApiExample)
- URLs remain stable across versions, which leads to:
   - Fewer changes
   - Less source control churn
   - HATEOAS, if every needed or desired

## Future Enhancements

After these changes are integrated, a few other enhancements are possible and potentially interesting:

- Adding the requested API version in instrumentation telemetry
- Discovery and reverse proxy mapping configuration from API version metadata (but no idea how - yet)

-----

cc: @captainsafia

### CODE REVIEW

File: src/Catalog.API/Catalog.API.http
Comment (captainsafia) [inline src/Catalog.API/Catalog.API.http:4]: Should we use a variable to capture the API version here?

```suggestion
GET {{Catalog.API_HostAddress}}/api/catalog/items{{targetVersion}}
```

File: tests/Catalog.FunctionalTests/CatalogApiTests.cs
Comment (captainsafia) [inline tests/Catalog.FunctionalTests/CatalogApiTests.cs:130]: Love this delta!

File: src/Webhooks.API/Program.cs
Comment (captainsafia) [inline src/Webhooks.API/Program.cs:8]: Is there a reason we have to pass the `IApiVersioningBuilder` via the arguments in this way? It seems to me that we could possible call `AddApiVersioning` inside the `AddDefaultOpenApi` call.

File: src/Ordering.API/Program.cs
Comment (captainsafia) [inline src/Ordering.API/Program.cs:18]: Nit: Can we keep the `UseDefaultOpenApi` call before the `MapX` calls for consistency?

File: src/Ordering.API/Program.cs
Comment (captainsafia) [inline src/Ordering.API/Program.cs:15]: Should we keep the `RequireAuthorization` call in `Program.cs`? It strikes me that it's a useful piece of information regardless of the version of the API used.

File: src/Catalog.API/Catalog.API.http
Comment (commonsensesoftware) [reply src/Catalog.API/Catalog.API.http:4]: We _can_. I followed the existing theme here. There wasn't a variable for `v{{1}}`, but there certainly can be. It _might_ convenient in this scenario to run each API. In a scenario the services evolve independently over time, it's more likely API versions will be heterogenous and, therefore, a single API version parameter is less useful.

It's a simple change that I'm happy to make if you'd like to see it added.

File: src/Webhooks.API/Program.cs
Comment (commonsensesoftware) [reply src/Webhooks.API/Program.cs:8]: The reason that `IApiVersioningBuilder` is passed here has to do with how API Versioning is configured. As an example, `IMvcCoreBuilder.AddApiExplorer` is different than `IApiVersioningBuilder.AddApiExplorer`, but they intentionally have the same name as they express the same intention. Under the hood, the later calls the former.

It would be possible to **not** pass the `IApiVersioningBuilder` and have a setup inside `AddDefaultOpenApi` that looks like:

```c#
builder.Services.AddApiVersioning().AddApiExplorer();
```

There's a few reasons why I chose not to do that:

1. Ordering can matter. If `AddApiVersioning()` hasn't already been called and configured, then services _might_ not be configured as expected. It's an edge case not likely to happen in a reference application, but someone copying and modifying it _could_ run into that problem in a very non-obvious way.
2. The name `AddDefaultOpenApi` implies that is configuring OpenAPI. If it also adds API Versioning, that is neither self-describing nor intrinsically obvious.
3. `AddDefaultOpenApi` is a shared extension method. It's possible that different APIs might configure API Versioning different ways; for example, different styles, different defaults, etc. It's better to the keep the API Versioning configuration separate so that it is obvious what and when it is configured. The API Versioning extensions for OpenAPI tend to be the same or overlap for all APIs so keeping it inside `AddDefaultOpenApi` makes sense. It may not even be invoked. By passing `IApiVersioningBuilder`, the ordering problem is solved because you had to have called `AddApiVersioning()` at least once.

I'll let you cogitate on that and decide if you'd like to change it.

File: src/Ordering.API/Program.cs
Comment (commonsensesoftware) [reply src/Ordering.API/Program.cs:15]: This one I was less sure on. If I'm not mistaken, adding `RequireAuthorization` will recursively apply to all endpoints under the associated group; regardless, of the depth of the group. That seems straight forward and makes sense. I was wondering whether all versions of an API would use the same authorization. Probably? 🤷🏽‍♂️

If you don't see any other issues with having `RequireAuthorization` in `Program.cs`, I can revert it. A future scenario I considered that would fail in a not-so-obvious way was:

```c#
app.NewVersionedApi("Orders")
   .MapOrdersApiV1()        // ← no authz
   .MapOrdersApiV2()
   .RequireAuthorization(); // ← applies to V2 only
```

If the order is reversed, I _believe_ it will work:

```c#
app.NewVersionedApi("Orders") // ← empty, top-level group
   .RequireAuthorization()    // ← applies from top-level down, recursively
   .MapOrdersApiV1()          // ← v1 group under api with authz
   .MapOrdersApiV2();         // ← v2 group under api with authz
```

File: src/Ordering.API/Program.cs
Comment (commonsensesoftware) [reply src/Ordering.API/Program.cs:18]: Unfortunately - no. This _limitation_ was one of the most infuriating part of _Minimal APIs_ (for me) when they were introduced. I understand why it works the way it does, but when all of the pieces are put together, things only work if configured in a very specific order. The issues are a perfect storm of all the libraries involved.

### Issue 1

ASP.NET Core backs `WebApplication` with an `EndpointDataSource` that it puts into the DI container. This is great for deferred evaluation since the backig services are designed to handle changes to the underlying endpoint collections.

https://github.com/dotnet/aspnetcore/blob/dc1acba9cd1374a8a8560bee655682e1a72de3eb/src/Http/Routing/src/DependencyInjection/RoutingServiceCollectionExtensions.cs#L75

### Issue 2

The API Explorer extensions provided API Versioning define the `IApiVersionDescriptionProvider` service, which is responsible for collating all of the API versions in the application. If all of the `Endpoint`s are deprecated, it will also handle reporting that a given API version is deprecated. This service is used to enumerate all API versions in an application. A common usage scenario is dynamically creating an OpenAPI document for each API version.

The `IApiVersionDescriptionProvider` service is backed by a collection of `IApiVersionMetadataCollationProvider` services that know how to collate API versions across _Minimal APIs_ and good 'ol controllers. These are all in the DI container and backed by the registered `EndpointDataSource` and `IActionDescriptorCollectionProvider`, respectively. At this point, there are no `Endpoint`s registered with `EndpointDataSource`; therefore, the _Minimal API_ collator doesn't discover anything. These services are aware that the underlying sources can change and will adapt, but the changes have to happen first.

https://github.com/dotnet/aspnet-api-versioning/blob/22a775b8eae36147587faaa062d7555260c94736/src/AspNetCore/WebApi/src/Asp.Versioning.Mvc.ApiExplorer/IEndpointRouteBuilderExtensions.cs#L27

### Issue 3

When you finally call `UseSwaggerUI()`, Swashbuckle uses eager evaluation. It's building the middleware for the pipeline and the DI container has already been built, so it isn't that unreasonable.

https://github.com/domaindrivendev/Swashbuckle.AspNetCore/blob/8f363f7359cb1cb8fa5de5195ec6d97aefaa16b3/src/Swashbuckle.AspNetCore.SwaggerUI/SwaggerUIBuilderExtensions.cs#L34

As a result, when the setup for `SwaggerUIOptions` is evaluated in `UseSwaggerUI`:

1. there are no _Minimal APIs_ registered
2. API Versioning doesn't see or collate any API versions from _Minimal APIs_ - yet
3. one or more OpenAPI documents are not generated (which is wrong and unexpected)

-----

You might be wondering why this hasn't been a more prevalent issue. If you're **not** using API Versioning, then it's not a problem because there's only a single OpenAPI document or you've hard-coded the mapping in the configuration setup. Once you want to use the API Versioning API Explorer extensions to drive the configuration automatically from the application, things start to fall down if they are not done in the correct order.

I've had more than one customer file an issue about this behavior. I always direct them to the API Versioning example projects. The differences in the example setups versus the vanilla out-of-the-box templates are only meant to vary by what is absolutely necessary to get things running, not my personal dogma about how I think the template should be structured.

From my experience, combining all of these concepts together requires `UseDefaultOpenApi ` be the last thing or nearly the last thing in the configuration. It definitely needs to be after `MapXXX` calls because that is the only way for API Versioning to discover them (automatically).

We wont' solve the problem here, but I fully support making this experience better. The order should not really matter. This behavior is something you probably want to account for in the forthcoming OpenAPI extensions that are planned for .NET 9.

File: src/Ordering.API/Program.cs
Comment (commonsensesoftware) [reply src/Ordering.API/Program.cs:15]: Partially reverting this back. I think the way we _recommend_ this should be done is like this:

```c#
var orders = app.NewVersionedApi("Orders");

orders.MapOrdersApiV1()
      .RequireAuthorization();

// if 2.0 was added, this is what it would look like
orders.MapOrdersApiV2()
      .RequireAuthorization();
```

It's not a hard requirement, but that makes things clear and separated.

Comment (commonsensesoftware) [reply src/Ordering.API/Program.cs:15]: Partially reverting this back. I think we want to _recommend_ that it be configured like so:

```c#
var orders = app.NewVersionedApi("Orders");

orders.MapOrdersApiV1()
      .RequireAuthorization();

// if 2.0 is added, it would look like this
orders.MapOrdersApiV2()
      .RequireAuthorization();
```

This is not a hard requirement, but it keeps things separate and clear.

File: src/Catalog.API/Catalog.API.http
Comment (commonsensesoftware) [reply src/Catalog.API/Catalog.API.http]: I've added a parameter as requested

File: src/Webhooks.API/Program.cs
Comment (commonsensesoftware) [reply src/Webhooks.API/Program.cs]: I've restructured this just a bit to make it read more clearly and retain the original functionality.

### Adding OpenAPI with API Versioning

```c#
var withApiVersioning = builder.Services.AddApiVersioning();
builder.AddDefaultOpenApi(withApiVersioning);
```

### Adding OpenAPI without API Versioning (Original)

```c#
builder.AddDefaultOpenApi();
```

File: src/eShop.ServiceDefaults/ConfigureSwaggerOptions.cs
Comment (qin-guan) [inline src/eShop.ServiceDefaults/ConfigureSwaggerOptions.cs:64]: Hey! Spotted this PR while searching on how to implement versioning. I believe this should be `!=`?
```suggestion
                if (text[^1] != '.')
```

File: src/eShop.ServiceDefaults/ConfigureSwaggerOptions.cs
Comment (qin-guan) [inline src/eShop.ServiceDefaults/ConfigureSwaggerOptions.cs:102]: Small quality of life improvement to make the link clickable!
```suggestion
                    text.Append($"<{link.LinkTarget.OriginalString}>");
```

File: src/eShop.ServiceDefaults/ConfigureSwaggerOptions.cs
Comment (commonsensesoftware) [reply src/eShop.ServiceDefaults/ConfigureSwaggerOptions.cs]: Good catch. Fixed.

File: src/eShop.ServiceDefaults/ConfigureSwaggerOptions.cs
Comment (commonsensesoftware) [reply src/eShop.ServiceDefaults/ConfigureSwaggerOptions.cs:105]: Nice suggestion. I don't think adding `<` and `>` alone will do the trick; however, I've refactored the implementation to render an unordered list of links, if any. If any links for HTML pages are defined, it will render like so:

```html
<h4>Links</h4>
<ul>
 <li><a href="policy.html">Versioning Policy</a></li>
</ul>
```

The [API Versioning examples](https://github.com/dotnet/aspnet-api-versioning/tree/main/examples) for OpenAPI have been updated to reflect this change as well.

File: src/Webhooks.API/Program.cs
Comment (captainsafia) [reply src/Webhooks.API/Program.cs]: @commonsensesoftware Thanks for sharing this! I have a better understanding of the problem here after taking a look at what it would take to integrate Asp.Versioning with the new built-in support in Microsoft.AspNetCore.OpenApi. We can discuss this further next week but for now I appreciate why this is setup this way.

**Review by captainsafia — APPROVED** [review]
Approving for now.

Some of this code will need to change once we update eShop to use the new APIs in `Microsoft.AspNetCore.OpenApi` for OpenAPI document generation.

We can track the changes that will need to be made in a separate issue as part of the work.

Thanks for opening this PR and for your patience here, @commonsensesoftware!

**Comment by captainsafia** [comment]
@commonsensesoftware It looks like CI wasn't running on this before due to the merge conflicts. It looks like `ApiVersioningVersion` isn't explicit defined anywhere. I assume we'll want to target `8.1.0` for all packages?

**Comment by commonsensesoftware** [comment]
@captainsafia yes, it should be `8.1.0`. there was a merge conflict with the original submission and I rebased in the upstream. I goofed something up during the fix. I tested it ran as expected so I'm not sure how this happened. Something was cached I guess. I can add that in quickly if you like.

**Comment by commonsensesoftware** [comment]
Added. Hopefully the CI build runs and is successful this time.

**Pull request merged**
